src/Semantic/Api/Diffs.hs

```diff
@@ -157,51 +176,29 @@ instance (Foldable syntax, Functor syntax, HasDeclaration syntax) => SummarizeDi
         = TOCSummaryFile path language changes (V.cons (TOCSummaryError errorText (converting #? errorSpan)) errors)
 
 
-type DiffActions term diff =
-  ( Bifoldable diff
-  , DiffTerms term diff
-  , DOTGraphDiff diff
-  , JSONGraphDiff diff
-  , JSONTreeDiff diff
-  , SExprDiff diff
-  , ShowDiff diff
-  , LegacySummarizeDiff term diff
-  , SummarizeDiff term diff
-  )
-
-doDiff
-  :: DiffEffects sig m
-  => Decorate Loc ann
-  -> (forall term diff . DiffActions term diff => diff ann ann -> m output)
+diffWith
+  :: (forall term . c term => DiffTerms term, DiffEffects sig m)
+  => [(Language, SomeParser c Loc)]
+  -> (forall term . c term => DiffFor term Loc Loc -> m output)
   -> BlobPair
   -> m output
-doDiff decorate render blobPair = do
-  SomeTermPair terms <- doParse blobPair decorate
-  diff <- diffTerms blobPair terms
-  render diff
+diffWith parsers render blobPair = parsePairWith parsers (render <=< diffTerms blobPair) blobPair
+
+decoratingDiffWith
+  :: forall ann c output m sig
+  .  (forall term . c term => DiffTerms term, DiffEffects sig m)
+  => [(Language, SomeParser c Loc)]
+  -> (forall term . c term => Blob -> term Loc -> term ann)
+  -> (forall term . c term => DiffFor term ann ann -> m output)
+  -> BlobPair
+  -> m output
+decoratingDiffWith parsers decorate render blobPair = parsePairWith parsers (render <=< diffTerms blobPair . bimap (decorate blobL) (decorate blobR)) blobPair where
+  (blobL, blobR) = fromThese errorBlob errorBlob (runJoin blobPair)
+  errorBlob = Prelude.error "evaluating blob on absent side"
```

Comment:
Is it possible to throw either a fused-effects `Error` or, if this error is not recoverable, create an exception type and `throw` that?

Reply:
It’s impossible for this to happen; the shape of the `These` in the `BlobPair` determines the shape of the `These` that this is eliminating. Thus, we’re guaranteed to have a blob on a side if we have a term on that side, so this is actually total.

Reply:
Okay, cool, `error` is the right thing then.